syntax = "proto2";

package ChakraProtoMsg;

enum NodeType
{
  INVALID_NODE = 0;
  MEM_LOAD_NODE = 1;
  MEM_STORE_NODE = 2;
  COMP_NODE = 3;
  COMM_SEND_NODE = 4;
  COMM_RECV_NODE = 5;
  COMM_COLL_NODE = 6;
}

enum AttributeType
{
  FLOAT = 0;
  INT = 1;
  STRING = 2;
  FLOATS = 3;
  INTS = 4;
  STRINGS = 5;
}

enum CollectiveCommType
{
  ALL_REDUCE = 0;
  ALL_TO_ALL = 1;
  ALL_GATHER = 2;
  REDUCE_SCATTER = 3;
}

message AttributeProto {
  required string name = 1;
  required AttributeType type = 2;
  optional string doc_string = 3;
  optional float f = 4;
  optional int64 i = 5;
  optional string s = 6;
  repeated float floats = 7;
  repeated int64 ints = 8;
  repeated bytes strings = 9;
}

message Node {
  required uint64 id = 1;
  required string name = 2;
  required NodeType type = 3 [default = INVALID_NODE];
  repeated uint64 parent = 4;
  optional string inputs = 5;
  optional string input_shapes = 6;
  optional string input_types = 7;
  optional string outputs = 8;
  optional string output_shapes = 9;
  optional string output_types = 10;
  repeated AttributeProto attribute = 11;
}
